src/bit-systems/pdf-system.ts

```diff
@@ -0,0 +1,85 @@
+import { PDFResources } from "../inflators/pdf";
+import { JobRunner } from "../utils/coroutine-utils";
+import { EntityID } from "../utils/networking-types";
+import { disposeMaterial } from "../utils/three-utils";
+import { waitForMediaLoaded } from "./media-loading";
+
+// TODO This can go away if we make bit-components becomes typescript file
+export const PDFResourcesMap = (MediaPDF as any).map as Map<EntityID, PDFResources>;
+
+/**
+ * Warning! This require statement is fragile!
+ *
+ * How it works:
+ * require -> require the file after all import statements have been called, particularly the configs.js import which modifies __webpack_public_path__
+ * !! -> don't run any other loaders
+ * file-loader -> make webpack move the file into the dist directory and return the file path
+ * outputPath -> where to put the file
+ * name -> how to name the file
+ * Then the path to the worker script
+ */
+GlobalWorkerOptions.workerSrc =
+  require("!!file-loader?outputPath=assets/js&name=[name]-[hash].js!pdfjs-dist/build/pdf.worker.min.js").default;
+
+type Aspect = { width: number; height: number };
+
+export function* loadPageJob(
+  { pdf, canvasContext, material }: PDFResources,
+  pageNumber: number
+): Generator<any, Aspect, any> {
+  const page: PDFPageProxy = yield pdf.getPage(pageNumber);
+  const viewport = page!.getViewport({ scale: 3 });
+  (material.map!.image as HTMLCanvasElement).width = viewport.width;
+  (material.map!.image as HTMLCanvasElement).height = viewport.height;
+  const renderTask = page!.render({ canvasContext, viewport, intent: "print" });
+  yield renderTask.promise;
+  material.map!.needsUpdate = true;
+  material.needsUpdate = true;
+  return { width: viewport.width, height: viewport.height };
+}
+
+export function fitToAspect(o: Object3D, { width, height }: Aspect) {
+  // Define a constant c to preserve the original area (scale.x * scale*y)
+  const c = Math.sqrt((o.scale.x * o.scale.y) / (width * height));
+  o.scale.x = width * c;
+  o.scale.y = height * c;
+  o.matrixNeedsUpdate = true;
+}
+
+export function* loadPageAndSetScale(world: HubsWorld, eid: EntityID, pageNumber: number) {
+  const aspect = yield* loadPageJob(PDFResourcesMap.get(eid)!, pageNumber);
+  const mesh = world.eid2obj.get(eid)!;
+  // HACK The loading animation may still be controlling the pdf scale,
+  // so wait until it is finished. This is rarely necessary.
+  yield* waitForMediaLoaded(world, mesh.parent!.eid!);
+  fitToAspect(mesh, aspect);
+}
+
+const jobs = new JobRunner();
+const pdfQuery = defineQuery([MediaPDF, NetworkedPDF]);
+const pdfExitQuery = exitQuery(pdfQuery);
+export function pdfSystem(world: HubsWorld) {
+  pdfQuery(world).forEach(function (eid) {
+    if (MediaPDF.pageNumber[eid] !== NetworkedPDF.pageNumber[eid]) {
+      MediaPDF.pageNumber[eid] = NetworkedPDF.pageNumber[eid];
+
+      jobs.stop(eid);
+      jobs.add(eid, () => loadPageAndSetScale(world, eid, NetworkedPDF.pageNumber[eid]));
```

Comment:
We talked about all of this in a 1:1  but I'm leaving it here too for discussion. 

We are explicitly ticking every job in each system. My understanding is that if we ticked all the jobs at the same time that would still guarantee the execution in the same frame point as if we were ticking them individually. Based on that assumption the API I imagine would be quite similar to this but hiding the `jobId` map inside the jobs runner and ticking all jobs from there called from the main tick. Something like:
```javascript
const jobId = APP.jobRunner.add((clear: ClearFunction, abort: AbortSignal) => myJob(clear, abort));
...
APP.jobRunner.cancel(jobId);
```
or:
```javascript
const myJob = (clear: ClearFunction, abort: AbortSignal) => myJob(clear, abort);
APP.jobRunner.add(myJob);
...
APP.jobRunner.cancel(myJob);
```

Reply:
> that would still guarantee the execution in the same frame point as if we were ticking them individually.

Not quite -- If we tick all jobs (registered by any system) at some point in the frame, that point is when they'll run (not at the point when they were registered). 

That's probably fine, but introduces some additional indirection we'd have to keep in mind.

Reply:
Also, so far as we have used these we have almost always wanted to stop the job when the entity a job is "about" goes away. If job ids are assigned, we could make a separate map from entity to jobid, but that seems unnecessary.

Reply:
Got it, all that makes sense.